pkgs/development/tools/rust/cargo-lambda/default.nix

```diff
@@ -10,13 +10,13 @@
 
 rustPlatform.buildRustPackage rec {
   pname = "cargo-lambda";
-  version = "0.13.0";
+  version = "0.18.1";
 
   src = fetchFromGitHub {
     owner = pname;
     repo = pname;
     rev = "v${version}";
-    sha256 = "sha256-JwrTwCn4qXcjXjy4cPYM3WCblJS0J5Xxnc6WESMwCpU=";
+    sha256 = "sha256-un+GQflxhMHCMH5UEeUVsYx59ryn7MR4ApooeOuhccc=";
   };
 
   cargoSha256 = "sha256-p3q5S6IFQQgNp/MHGSUE1DVLFyMLWDTv/dxrUACKSWo=";
```

Comment:
@calavera @taylor1791 

How did we get this attribute in the first place?
Do we go in the cargo lambda project and generate the git hash on `Cargo.lock`?

I assume we need to regenerate it.

Reply:
Yes this needs to be regenerated. IIRC, I ran nix-build on this package and used the value in the error message.

Reply:
I haven't found anything I can get the hash from. Nothing is telling me what the new hash could be. Any ideas with this information?

This as far as I can get, wishing to become more helpful soon.

I also discovered this.
Today is April 6 so none of these are from the builds I've done.


<details>
<summary>`ls` nix store for cargo-lambda</summary>

```
[nix-shell:~/.cache/nixpkgs-review/rev-b7a1580b29ce12b4060f0fe4a768a726ea12615b-1]$ ll /nix/store/*cargo-lambda*
-rw------- 1 root root    0 Apr  5 17:52 /nix/store/05pcykb3lafvflsnp50xkm0d93zsijf5-cargo-lambda-0.18.1-vendor.tar.gz.lock
-r--r--r-- 1 root root 3513 Jan  1  1970 /nix/store/07gk1av9h3pcznmdvfyg62ia7qbmjzph-cargo-lambda-0.18.1.drv
-r--r--r-- 1 root root 4093 Jan  1  1970 /nix/store/08srishyd7blg0w08m7acdwqdxscc6la-cargo-lambda-0.18.1-vendor.tar.gz.drv
-r--r--r-- 1 root root 4093 Jan  1  1970 /nix/store/8x0szq1qp0lll9fazxwskh2inmhiwi4s-cargo-lambda-0.13.0-vendor.tar.gz.drv
-r--r--r-- 1 root root 3513 Jan  1  1970 /nix/store/8x8pgga2hb1jkjyjhcij8m2yn9i1aa5p-cargo-lambda-0.13.0.drv
-r--r--r-- 1 root root 3369 Jan  1  1970 /nix/store/ahpp6lzn3dz1a3hx5v4f690nds0a4ahy-cargo-lambda-0.18.1.drv
-rw------- 1 root root    0 Apr  5 22:01 /nix/store/alwr2c920b66p46qhfki4c10xrllrf4r-cargo-lambda-0.18.1-vendor.tar.gz.lock
-r--r--r-- 1 root root 4093 Jan  1  1970 /nix/store/cyjsrhc7xm705xmny7kkwi3v2y55pg0f-cargo-lambda-0.18.1-vendor.tar.gz.drv
-r--r--r-- 1 root root 4093 Jan  1  1970 /nix/store/l0p8mchk29666apdy106p38fxn4d5azj-cargo-lambda-0.18.1-vendor.tar.gz.drv
-r--r--r-- 1 root root 3513 Jan  1  1970 /nix/store/p5pwzwpsbkkx1c0rh0bls35xbpjlgyh4-cargo-lambda-0.18.1.drv

/nix/store/zsnx8h1aa8csnri0d5q5sqb6hva8ym89-cargo-lambda-0.13.0:
total 476
dr-xr-xr-x   3 root root     4096 Jan  1  1970 ./
drwxrwxr-t 629 root nixbld 475136 Apr  6 21:38 ../
dr-xr-xr-x   2 root root     4096 Jan  1  1970 bin/

```
</details>

<details>
<summary>Build + Version check</summary>

```
nixpkgs git:(update/cargo-lambda) nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0
$ git worktree add /home/wayne/.cache/nixpkgs-review/rev-b7a1580b29ce12b4060f0fe4a768a726ea12615b/nixpkgs 0a5c42df2a9bdee2814761565b15b54873340e27
Preparing worktree (detached HEAD 0a5c42df2a9)
Updating files: 100% (34361/34361), done.
HEAD is now at 0a5c42df2a9 Merge pull request #223835 from maxammann/fix-tracy-macos
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f /home/wayne/.cache/nixpkgs-review/rev-b7a1580b29ce12b4060f0fe4a768a726ea12615b/nixpkgs -qaP --xml --out-path --show-trace --no-allow-import-from-derivation
$ git merge --no-commit --no-ff b7a1580b29ce12b4060f0fe4a768a726ea12615b
Automatic merge went well; stopped before committing as requested
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f /home/wayne/.cache/nixpkgs-review/rev-b7a1580b29ce12b4060f0fe4a768a726ea12615b/nixpkgs -qaP --xml --out-path --show-trace --no-allow-import-from-derivation --meta
1 package updated:
cargo-lambda (0.13.0 → 0.18.1)

$ nix build --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/wayne/.cache/nixpkgs-review/rev-b7a1580b29ce12b4060f0fe4a768a726ea12615b/build.nix
error: builder for '/nix/store/l0p8mchk29666apdy106p38fxn4d5azj-cargo-lambda-0.18.1-vendor.tar.gz.drv' failed with exit code 1;
       last 10 log lines:
       > source = "git+https://github.com/rust-lang/cargo#c6c69cde0dc92cecc1f92d03b92dc7cd47d8e6a2"
       > source = "git+https://github.com/rust-lang/cargo#c6c69cde0dc92cecc1f92d03b92dc7cd47d8e6a2"
       > source = "git+https://github.com/rust-lang/cargo#c6c69cde0dc92cecc1f92d03b92dc7cd47d8e6a2"
       > source = "git+https://github.com/rust-lang/cargo#c6c69cde0dc92cecc1f92d03b92dc7cd47d8e6a2"
       >
       > ERROR: The Cargo.lock contains git dependencies
       >
       > This is currently not supported in the fixed-output derivation fetcher.
       > Use cargoLock.lockFile / importCargoLock instead.
       >
       For full logs, run 'nix log /nix/store/l0p8mchk29666apdy106p38fxn4d5azj-cargo-lambda-0.18.1-vendor.tar.gz.drv'.
error: 1 dependencies of derivation '/nix/store/ahpp6lzn3dz1a3hx5v4f690nds0a4ahy-cargo-lambda-0.18.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/0bvmm04136ivg916h9zf11gn35b55b76-review-shell.drv' failed to build
1 package failed to build:
cargo-lambda

error: build log of '/nix/store/ahpp6lzn3dz1a3hx5v4f690nds0a4ahy-cargo-lambda-0.18.1.drv!*' is not available
error: build log of '/nix/store/mssk412yvxbrygggxw4hn4qziy07p09n-cargo-lambda-0.18.1' is not available
$ /nix/store/0sxbaj71c4c4n43qhdxm31f56gjalksw-nix-2.13.3/bin/nix-shell /home/wayne/.cache/nixpkgs-review/rev-b7a1580b29ce12b4060f0fe4a768a726ea12615b/shell.nix
```
</details>

Reply:
I didn't realize a git dependency was added. If you replace `cargoSha256` with the following and cargo-lambda builds again.

```nix
  cargoLock = {
    lockFile = "${src}/Cargo.lock";

    outputHashes = {
      "cargo-test-macro-0.1.0" = "sha256-XvTKAbP/r1BthpEM84CYZ2yfJczxqzscGkN4JXLgvfA=";
    };
  };
```

However, the test phase is failing with the following error. I tried a few things and I ran out of time to try and fix it today. 
```
 run pkg_config fail: "Could not run `\"pkg-config\" \"--libs\" \"--cflags\" \"openssl\"`\nThe pkg-config command could not be found.\n\nMost likely, you need to install a pkg-config package for your OS.\nTry `apt install pkg-config`, or `yum install pkg-config`,\nor `pkg install pkg-config`, or `apk add pkgconfig` depending on your distribution.\n\nIf you've already installed it, ensure the pkg-config command is one of the\ndirectories in the PATH environment variable.\n\nIf you did not expect this build to link to a pre-installed system library,\nthen check documentation of the openssl-sys crate for an option to\nbuild the library from source, or disable features or dependencies\nthat require pkg-config."
```

Along the way I set the hashes to `""` and received errors like this.
```
error: hash mismatch in fixed-output derivation '/nix/store/gg20b50c9jxfy4iqivmzm85gl7igvm8k-source.drv':
         specified: sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=
            got:    sha256-un+GQflxhMHCMH5UEeUVsYx59ryn7MR4ApooeOuhccc=
error: 1 dependencies of derivation '/nix/store/xcqplai1jqxwjcw9ax9v4w8fbc9y91jy-cargo-lambda-0.18.1.drv' failed to bui
```

Regarding downloading dev dependencies. One obvious reason is because it runs the tests.

Regarding the static binaries. I didn't use them because I was pattern matching. For example [cargo-hack](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/rust/cargo-hack/default.nix) provides [static binaries](https://github.com/taiki-e/cargo-hack/releases), although not as complete as [cargo-lambda](https://github.com/cargo-lambda/cargo-lambda/releases), and they still build from source. Generally speaking, I get the impression that nixpkgs prefers to build from source when possible. I don't want to presume the reasons why.

Reply:
@calavera `buildRustPackage` runs `cargo test` in the [check phase](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/hooks/cargo-check-hook.sh#L36). Cargo assumes it needs all of the development dependencies to run tests.

Reply:
@calavera I think you misunderstand how nix works. The packages are built on a build server. Everyone installs the binaries produced by the build server when they install any package. No end user runs test or compiles anything.

Should we switch to the precompiled binaries? I don't know, but I'll see if I can find someone who knows more than I.

Reply:
While formulating my question I noticed that there is no pre-compiled static binary for aarch64-darwin, there is only one for x86_64-darwin. It does look like there is a whl though. Is creating one an option?

If a pre-compiled version for aarch64-darwin was available, I don't know if that is the preferred approach. I am asking the community [here](https://discourse.nixos.org/t/when-to-use-precompiled-binaries-in-a-nixpkgs-derivation/27087).

Reply:
@calavera A few members of the community responded. This answer in particular made made me think about nixpkgs differently. There were a few more tactical answers about supported architectures and trust that were also valuable.

> [nixpkgs] as an ecosystem, and not a package manger, you can do thing like change version of rust compiler, change version of llvm that rust uses and so on, and so you would expect that rust crates are all using/build with versions of libraries from nixpkgs

Do they convince you we shouldn't use the pre-compiled static binaries?

@waynevanson I fixed the failing tests and disabled a few new ones that reach out to the network. I also switched out the IFD that @alyssais mentioned. I pushed these changes to [taylor1791/nixpkgs!update-cargo-lambda](https://github.com/taylor1791/nixpkgs/commit/0289762e1d7aae3f45a759c3545b5b5d74d2736d). Given the work you have done, it feels appropriate for you to receive credit. If you agree, feel free to incorporate the changes from the branch I pushed.

Reply:
@taylor1791 
Great work! I've committed your changes here.
I haven't done that much, I just think it's better to keep all this information in one place.